concourse/steps/release.py

```diff
@@ -390,17 +407,49 @@ def validate(self):
     def apply(
         self,
     ):
+        # clean repository if required
+        worktree_dirty = bool(self.github_helper._changed_file_paths())
+        if worktree_dirty:
+            self.github_helper._stash_changes()
+
+        # update version file
         next_cycle_dev_version = self._calculate_next_cycle_dev_version(
             release_version=self.release_version,
             version_operation=self.version_operation,
             prerelease_suffix=self.prerelease_suffix,
         )
-        # Prepare version file for next dev cycle
-        self.github_helper.create_or_update_file(
-            file_path=self.repository_version_file_path,
-            file_contents=next_cycle_dev_version,
-            commit_message=f'Prepare next dev cycle {next_cycle_dev_version}'
-        )
+
+        # update version file
+        with open(self.repository_version_file_path, 'w') as f:
+            f.write(next_cycle_dev_version)
+
+        try:
+            # call optional dev cycle commit callback
+            if self.dev_cycle_commit_callback:
+                _invoke_callback(
+                    callback_script_path=self.dev_cycle_commit_callback,
+                    repo_dir=self.repo_dir,
+                    effective_version=next_cycle_dev_version,
+                )
+
+            dev_cycle_commit = self.git_helper.index_to_commit(
+                message=f'Prepare next dev cycle {next_cycle_dev_version}',
+            )
+
+            dev_cycle_commit_sha = dev_cycle_commit.hexsha
+
+            # forward head to new commit
+            self.github_helper.repo.head.set_commit(dev_cycle_commit_sha)
+
+            # Push release commit to remote
+            self.github_helper.push(
+                from_ref=dev_cycle_commit_sha,
+                to_ref=self.github_helper.default_branch,
+                use_ssh=True,
+            )
+        finally:
+            if worktree_dirty and self.github_helper._has_stash():
```

Comment:
git_helper
yes, both our API is confusing in that regard AND it sucks not to have static typing :-/